### PR TITLE
imap to six / replace Variable(new VariableImpl) with make_variable

### PR DESCRIPTION
**Pull request by ezyang**
CC @apaszke @colesbury

### CODE REVIEW

File: torch/csrc/autograd/variable.h [hunk 1]
```diff
@@ -62,9 +62,6 @@ struct Variable : public at::Tensor {
 
   inline Variable & operator=(Variable && rhs) &;
   inline Variable & operator=(const Variable & rhs) &;
-
-  // implicit conversion to Tensor
-  operator Tensor() const { return Tensor(pImpl, true); }
```
Comment (colesbury) [inline torch/csrc/autograd/variable.h]: 👍

File: torch/csrc/autograd/variable.h [hunk 2]
```diff
@@ -110,6 +107,19 @@ struct VariableImpl : public at::TensorImpl {
   friend struct VariableType;
 };
 
+inline Variable make_variable(at::Tensor data) {
+  return Variable(new VariableImpl(data), false);
+}
+
+inline Variable make_variable(at::Tensor data, std::shared_ptr<Function> grad_fn) {
+  return Variable(new VariableImpl(data, grad_fn), false);
```
Comment (colesbury) [inline torch/csrc/autograd/variable.h]: `new VariableImpl(std::move(data), std::move(grad_fn))` (and same with other `make_variable` overloads)

**Review by colesbury — COMMENTED** [review]
Looks good

File: torch/csrc/autograd/variable.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/autograd/variable.h]: I'm going to make these take const references instead. One less move.

Comment (colesbury) [reply torch/csrc/autograd/variable.h]: I don't think that's what you want. That will still invoke the copy constructor. 

For example, this code shouldn't result in extra `retain()` calls:

`make_variable(thing_that_returns_tensor())`

Comment (ezyang) [reply torch/csrc/autograd/variable.h]: OK, I guess then we would also need the move version... I'll make a value constructor then.

**Review by apaszke — APPROVED** [review]

**Pull request merged**